### PR TITLE
add exp,sqrt and rsqrt operations.

### PR DESCRIPTION
**Pull request by YZY010203**

### CODE REVIEW

File: cc/modules/protocol/mpc/snn/src/impl/exp.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+{
+    const int exp_iterations = 8;
+    vector<mpc_t> n(size,FloatToMpcType(1/(2^exp_iterations)));//n为实际迭代次数
+    GetMpcOpInner(Reconstruct2PC)->Run(n, n.size(), "1/n");
+    vector<mpc_t> m(size , 0) ; //
+    vector<mpc_t> result(size , 0);
+    const vector<mpc_t> one(size,FloatToMpcType(1));
+    GetMpcOpInner(DotProduct)->Run(a, n, m, size);//m=a/2的exp_iterations次方
+    
+    GetMpcOpInner(Reconstruct2PC)->Run(m, m.size(), "a/n");
```
Comment (vingstar) [inline cc/modules/protocol/mpc/snn/src/impl/exp.cpp:36]: No debugging info in  final code

File: cc/modules/protocol/public/include/protocol_ops.h [hunk 2]
```diff
@@ -371,36 +392,13 @@ class ProtocolOps {
...
 
-  // get \sqrt{a}
-  virtual int Sqrt(
-    const vector<string>& a,
-    vector<string>& output,
-    const attr_type* attr_info = nullptr) {
-    THROW_NOT_IMPL;
-  }
-
-  // get 1/\sqrt{a}
```
Comment (vingstar) [inline cc/modules/protocol/public/include/protocol_ops.h:382]: why these comments are removed?

File: cc/modules/protocol/mpc/snn/src/impl/exp.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    if (partyNum == PARTY_A) 
+        addVectors<mpc_t>(m, one, result, size);//result = 1+m
+    if(partyNum == PARTY_B)
+        result = m;
+    GetMpcOpInner(Reconstruct2PC)->Run(result, result.size(), "1+a/n");
+        
+    for(int i=0;i<exp_iterations;++i)
+        GetMpcOpInner(Square)->Run(result,result,size);
+    b = result;//result = (1 + x / n) ^ n,其中n = 2的8次方
```
Comment (vingstar) [inline cc/modules/protocol/mpc/snn/src/impl/exp.cpp:45]: Not using  Chinese comments

File: python/latticex/rosetta/secure/decorator/test_cases/test.sh [hunk 4]
```diff
@@ -51,9 +51,9 @@ echo "run standalone test case..."
 # test_op tf_const
 # test_op axis_test
 # test_op abs_test
-test_op save_v2
-test_op restore_v2
-test_op nn_ops_test
+#test_op save_v2
```
Comment (vingstar) [inline python/latticex/rosetta/secure/decorator/test_cases/test.sh:54]: better not affecting current other's unit tests

**Review by vingstar — CHANGES_REQUESTED** [review]
besides these detailed flaws indicated in in-line comments, more docs on algorithm idea description should be added.
Usage of these new OP should be added in API doc. 

And  I did NOT  see the replacement of native `tf.exp`  in the code.

File: cc/modules/protocol/mpc/snn/src/impl/exp.cpp (hunk 1 above)
Comment (YZY010203) [reply cc/modules/protocol/mpc/snn/src/impl/exp.cpp:36]: These lines have been deleted.

File: cc/modules/protocol/public/include/protocol_ops.h (hunk 2 above)
Comment (YZY010203) [reply cc/modules/protocol/public/include/protocol_ops.h:382]: I will  add them again and described accurately.

File: cc/modules/protocol/mpc/snn/src/impl/exp.cpp (hunk 3 above)
Comment (YZY010203) [reply cc/modules/protocol/mpc/snn/src/impl/exp.cpp:45]: OK.

File: cc/modules/protocol/mpc/snn/src/impl/exp.cpp (hunk 1 above)
Comment (YZY010203) [reply cc/modules/protocol/mpc/snn/src/impl/exp.cpp:36]: These lines have been deleted.

**Comment by vingstar** [comment]
Well Done! LGTM.

**Comment by vingstar** [comment]
Generally this  looks good, and I will fix some tiny issues.

File: python/latticex/rosetta/rtt/ops/rtt_math_ops.py [hunk 5]
```diff
@@ -354,6 +373,8 @@ def static_override_tf_ops_to_rtt_ops():
     tf.matmul = rtt_matmul
     tf.square = rtt_square
     tf.pow = rtt_pow
+    tf.sqrt = rtt_sqrt
+    tf.exp = rtt_exp
```
Comment (vingstar) [inline python/latticex/rosetta/rtt/ops/rtt_math_ops.py:377]: there is also a `tf.rsqrt` OP.

File: python/latticex/rosetta/secure/decorator/test_cases/rsqrt.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+#
+init = tf.compat.v1.global_variables_initializer()
+sess = tf.compat.v1.Session()
+sess.run(init)
+print("input xa:", sess.run(xa))
+
+###########
+print("=========================== tf op rsqrt 1")
+xc = tf.reciprocal(tf.sqrt(xa))
```
Comment (vingstar) [inline python/latticex/rosetta/secure/decorator/test_cases/rsqrt.py:24]: please test with `tf.rsqrt` directly.

File: python/latticex/rosetta/secure/decorator/test_cases/test.sh [hunk 7]
```diff
@@ -63,13 +63,16 @@ test_op nn_ops_test
 # test_op add
 # test_op sub
 # test_op mul
-test_op division_test
+#test_op division_test
```
Comment (vingstar) [inline python/latticex/rosetta/secure/decorator/test_cases/test.sh:66]: enable all these tests

**Comment by YZY010203** [comment]
Do I need to change these issues and submit them again？

**Comment by vingstar** [comment]
> Do I need to change these issues and submit them again？

yep, it will be great!

**Comment by YZY010203** [comment]
I have fix the problems.

**Comment by vingstar** [comment]
@YZY010203  
thanks for contributing! 
LGTM, and I will fix some tiny style issues later.

**Pull request merged**